python/ray/serve/_private/application_state.py

```diff
@@ -353,33 +358,32 @@ def deploy_config(
         """
 
         self._deployment_timestamp = deployment_time
-        self._set_target_state_config(config)
 
         config_version = get_app_code_version(config)
         if config_version == self._target_state.code_version:
             try:
                 overrided_infos = override_deployment_info(
                     self._name,
                     self._target_state.deployment_infos,
-                    self._target_state.config,
+                    config,
                 )
                 self._check_routes(overrided_infos)
-                self._set_target_state_deployment_infos(overrided_infos)
-            except (TypeError, ValueError, RayServeException):
                 self._set_target_state(
-                    deployment_infos=None,
-                    code_version=None,
-                    target_config=self._target_state.config,
+                    # Code version doesn't change.
+                    code_version=self._target_state.code_version,
+                    # Everything else must reflect the new config.
+                    deployment_infos=overrided_infos,
+                    target_config=config,
+                    target_capacity=target_capacity,
+                    target_capacity_direction=target_capacity_direction,
                 )
+            except (TypeError, ValueError, RayServeException):
+                self._clear_target_state()
                 self._update_status(
                     ApplicationStatus.DEPLOY_FAILED, traceback.format_exc()
                 )
             except Exception:
-                self._set_target_state(
-                    deployment_infos=None,
-                    code_version=None,
-                    target_config=self._target_state.config,
-                )
+                self._clear_target_state()
```

Comment:
This appears to be a behavior change -- previously we set `target_config=self._target_state.config` and now it's cleared to `None`. What's the purpose for this change and are you sure it doesn't have side effects?

Reply:
Good catch, I didn't see a reason that we were setting `target_config` to `self._target_state.config`, and the reason for these calls seemed to be to clear the target state altogether.

I tried reading through the code, but I didn't see any reason to keep the `self._target_state.config`, and all unit tests continue to pass even when I remove it.

@zcin could you confirm whether this behavior change is okay?

Reply:
I don't think this currently has any side effects. However in the long term instead of storing the app configs in the [controller](https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/controller.py#L982-L988), I think it's probably better to fetch them from the application state manager. The target state is the place that we store the full config for an application, so would it be better if we don't clear the target config?

Reply:
Makes sense. I updated the logic to store the config and clear the rest of the target state.